sktime/annotation/igts.py

```diff
@@ -0,0 +1,486 @@
+            and value series in columns.
+
+        change_points: list of ints
+            Locations of change points as integer indexes. By convention change points
+            include the identity segmentation, i.e. first and last index + 1 values.
+
+        Returns
+        -------
+        information_gain: float
+            Information gain score for the segmentation corresponding to the change
+            points.
+        """
+        segment_entropies = [
+            seg.shape[0] * entropy(seg) for seg in generate_segments(X, change_points)
+        ]
+        return entropy(X) - sum(segment_entropies) / X.shape[0]
+
+    def find_change_points(self, X: npt.ArrayLike) -> List[int]:
+        """Find change points.
+
+        Using a top-down search method, iteratively identify at most
+        `k_max` change points that increase the information gain score
+        the most.
+
+        Parameters
+        ----------
+        X: array_like
+            Time series data as a 2D numpy array with sequence index along rows
+            and value series in columns.
+
+        Returns
+        -------
+        change_points: list of ints
+            Locations of change points as integer indexes. By convention change points
+            include the identity segmentation, i.e. first and last index + 1 values.
+        """
+        n_samples, n_series = X.shape
+        if n_series == 1:
+            raise ValueError(
+                "Detected univariate series, IGTS will not work properly"
+                " in this case. Consider augmenting your series to multivariate."
+            )
+        self.intermediate_results_ = []
+
+        # by convention initialize with the identity segmentation
+        current_change_points = self.identity(X)
+
+        for k in range(self.k_max):
+            ig_max = 0
+            # find a point which maximizes score
+            for candidate in self.get_candidates(n_samples, current_change_points):
+                try_change_points = {candidate}
+                try_change_points.update(current_change_points)
+                try_change_points = sorted(try_change_points)
+                ig = self.information_gain_score(X, try_change_points)
+                # logger.info(f"{ig=:.5f} for {candidate=} with {try_change_points=}")
+                if ig > ig_max:
+                    ig_max = ig
+                    best_candidate = candidate
+
+            current_change_points.append(best_candidate)
+            current_change_points.sort()
+            self.intermediate_results_.append(
+                ChangePointResult(
+                    k=k, score=ig_max, change_points=current_change_points.copy()
+                )
+            )
+
+            # logger.info(
+            #     f"BEST {ig_max=:.5f} for {best_candidate=} "
+            #     + f"current {current_change_points=}"
+            # )
+        return current_change_points
+
+
+class SegmentationMixin:
```

Comment:
should this not go into another file? Do you think its useful to be used by other algos?

Reply:
Yes, that would be ultimate goal, however I would like to keep this PR small, and work on the changes in following PRs.